include/anyode/anyode_numpy.hpp

```diff
@@ -1,308 +1,439 @@
+	const static NPY_TYPES float_type_tag = npy_real_type<Real_t>::type_tag;
+
+	Index_t get_ny() const
+
+	override {
+		return ny;
+	}
+
+	int get_mlower() const
+
+	override {
+		return mlower;
+	}
+
+	int get_mupper() const
+
+	override {
+		return mupper;
+	}
+
+	Index_t get_nnz() const
+
+	override {
+		return nnz;
+	}
+
+	int get_nquads() const
+
+	override {
+		return nquads;
+	}
+
+	int get_nroots() const
+
+	override {
+		return nroots;
+	}
+
+	Real_t get_dx0(Real_t t, const Real_t * const y)
+
+	override {
+		if (py_dx0cb == nullptr || py_dx0cb == Py_None) {
+			return this->default_dx0;
+		}
+		npy_intp dims[1] { static_cast<npy_intp>(this->ny) };
+		PyObject *py_yarr = PyArray_SimpleNewFromData(1, dims,
+				this->float_type_tag,
+				static_cast<void *>(const_cast<Real_t *>(y)));
+		PyArray_CLEARFLAGS(reinterpret_cast<PyArrayObject*>(py_yarr),
+				NPY_ARRAY_WRITEABLE);  // make yarr read-only
+		PyObject *py_arglist = Py_BuildValue("(dO)", (double) t, py_yarr);
+		PyObject *py_result = PyEval_CallObjectWithKeywords(this->py_dx0cb,
+				py_arglist, this->py_kwargs);
+		Py_DECREF(py_arglist);
+		Py_DECREF(py_yarr);
+		if (py_result == nullptr) {
+			throw std::runtime_error("get_dx0 failed (dx0cb failed)");
+		}
+		Real_t res = (Real_t) PyFloat_AsDouble(py_result);
+		Py_DECREF(py_result);
+		if ((
+
+		PyErr_Occurred()
+
+		) && (res == -1.0)) {throw std::runtime_error("get_dx0 failed (value returned by dx0cb could not be converted to float)");
+	}
+		return res;
+	}
+
+	Real_t get_dx_max(Real_t t, const Real_t * const y)
+
+	override {
+		if (py_dx_max_cb == nullptr || py_dx_max_cb == Py_None) {
+			return INFINITY;
+		}
+		npy_intp dims[1] { static_cast<npy_intp>(this->ny) };
+		PyObject *py_yarr = PyArray_SimpleNewFromData(1, dims,
+				this->float_type_tag,
+				static_cast<void *>(const_cast<Real_t *>(y)));
+		PyArray_CLEARFLAGS(reinterpret_cast<PyArrayObject*>(py_yarr),
+				NPY_ARRAY_WRITEABLE);  // make yarr read-only
+		PyObject *py_arglist = Py_BuildValue("(dO)", (double) t, py_yarr);
+		PyObject *py_result = PyEval_CallObjectWithKeywords(this->py_dx_max_cb,
+				py_arglist, this->py_kwargs);
+		Py_DECREF(py_arglist);
+		Py_DECREF(py_yarr);
+		if (py_result == nullptr) {
+			throw std::runtime_error("get_dx_max failed (dx_max_cb failed)");
+		}
+		Real_t res = (Real_t) PyFloat_AsDouble(py_result);
+		Py_DECREF(py_result);
+		if (
+
+		PyErr_Occurred() &&
+
+		(res == -1.0)) {
+			throw std::runtime_error(
+					"get_dx_max failed (value returned by dx_max_cb could not be converted to float)");
+		}
+		return res;
+	}
+	Status handle_status_(PyObject * py_result, const std::string what_arg) {
+		if (py_result == nullptr) {
+			throw std::runtime_error(what_arg + " failed");
+		} else if (py_result == Py_None) {
+			Py_DECREF(py_result);
+			return AnyODE::Status::success;
+		}
+		long result = PyInt_AsLong(py_result);
+		Py_DECREF(py_result);
+
+		if ((
+
```

Comment:
when indentation changes it's almost impossible to spot any changes to the logic.
here it looks like your IDE/code-formater messed up? (it looks like that if statement spans many rows now)
I'm open to having a clang format file check in to reduce noise from white space changes (we could e.g. take the one from symengine & their script bin/test_format_local.sh)

Reply:
Yes I think it screwed with the formatting. As for everything else: makes sense